### PR TITLE
fix: robust point location for degenerate configurations

### PR DESCRIPTION
**Pull request by sframba**
This PR adds a new method to determine if a point lies inside or outsilde a (convex) polyhedral cell.
Compared to the current method, the new implementation is slower, but handles gracefully the degenerate conditions of points lying on mesh vertices, edges and faces. It does so by:
- Triangulating each face of a cell in a consistent manner, by always choosing the lowest global-id vertex as root;
- Ordering each triangle again using global ids, so that the edges and vertices are encountered in a consistent order across MPI ranks;
- Computing the winding number (an integer) subtended by each triangle;
- If the point lies on a vertex, edge or face, this will be found consistently by all ranks, whatever the cell from which the face/edge/vertex is seen. This is true because the order in which the vertices are traversed (and thus the result of the calculation) is exactly the same no matter the originating cell or MPI rank;
- Finally, if a degenerate condition is found, the {face,edge,vertex}-to-element maps of the host element are used to choose consistently a single representative element. The adjacent element with the lowest global ID is chosen, ensuring general consistency across MPI ranks.

The only ambiguity left is when the point lies on a face/edge/vertex that is a boundary for an MPI rank, but an interior face/edge/vertex for another rank. In this case, both ranks will claim the point. However, at least one of them will consider the point to belong to a ghost element. This is a desirable outcome, since the desired behaviour in this case can be different according to the specific application (e.g., sources vs. receivers for the seismic wave case). This outcome leaves the choice to the caller method on how to handle this case, using ghost rank information.

The new point location algorithm is used for wave propagation kernels, fixing [issue 3104](https://github.com/GEOS-DEV/GEOS/issues/3104) and [issue 2888](https://github.com/GEOS-DEV/GEOS/issues/2888).

### CODE REVIEW

**Review by acitrain — APPROVED** [review]

File: src/coreComponents/mesh/utilities/ComputationalGeometry.hpp [hunk 1]
```diff
@@ -435,6 +434,375 @@ bool isPointInsidePolyhedron( arrayView2d< real64 const, nodes::REFERENCE_POSITI
...
+ * @param[in] ay y-coordinate of the first vertex
+ * @param[in] az z-coordinate of the first vertex
+ * @param[in] bx x-coordinate of the second vertex
+ * @param[in] by y-coordinate of the second vertex
+ * @param[in] bz z-coordinate of the second vertex
+ * @return 0 if the vertices coincide, +1 if a > b and -1 if b > a
+ */
+template< typename COORD_TYPE, typename POINT_TYPE >
+GEOS_HOST_DEVICE
+int lexicographicalCompareVertex( POINT_TYPE const ax, POINT_TYPE const ay, POINT_TYPE const az,
```
Comment (untereiner) [inline src/coreComponents/mesh/utilities/ComputationalGeometry.hpp:451]: (ax,ay,az) and (bx,by,bz) can be of different concrete types ? (with the assumption that these two types overload comparison operator)

File: src/coreComponents/mesh/utilities/ComputationalGeometry.hpp [hunk 2]
```diff
@@ -435,6 +434,375 @@ bool isPointInsidePolyhedron( arrayView2d< real64 const, nodes::REFERENCE_POSITI
...
+                    ( v3x * v1y - v3y * v1x ) * v2z;
+  if( sign > 0 )
+    return 1;
+  else if( sign < 0 )
+    return -1;
+  return 0;
+}
+/**
+ * @brief Method to find the reference element touching a vertex. The element with the lowest global ID is chosen
+ * list.
```
Comment (untereiner) [inline src/coreComponents/mesh/utilities/ComputationalGeometry.hpp]: "in chosen list" ?

File: src/coreComponents/mesh/utilities/ComputationalGeometry.hpp [hunk 3]
```diff
@@ -435,6 +434,375 @@ bool isPointInsidePolyhedron( arrayView2d< real64 const, nodes::REFERENCE_POSITI
...
+ * @param[in] nodesToElements map from nodes to elements
+ * @param[in] nodeLocalToGlobal global indices of nodes
+ * @param[in] elementLocalToGlobal global indices of elements
+ * @param[in] elemCenter coordinates of the element centroid
+ * @param[in] point coordinates of the query point
+ * @return whether the point is inside
+ */
+template< typename COORD_TYPE, typename POINT_TYPE >
+GEOS_HOST_DEVICE
+bool isPointInsideConvexPolyhedronRobust( localIndex element,
```
Comment (untereiner) [inline src/coreComponents/mesh/utilities/ComputationalGeometry.hpp]: is it possible to have two functions:
`integer computeWindingNumber(...)` and 
```
bool isPointInsideConvexPolyhedronRobust() 
{
 return computeWindingNumber() != 0;
}
 ```

**Review by untereiner — COMMENTED** [review]
Is it possible to add a test into` testComputationalGeometry.cpp` ?

File: src/coreComponents/mesh/utilities/ComputationalGeometry.hpp (hunk 2 above)
Comment (sframba) [reply src/coreComponents/mesh/utilities/ComputationalGeometry.hpp]: It was supposed to be `is chosen from the list`. I have corrected the comment

File: src/coreComponents/mesh/utilities/ComputationalGeometry.hpp (hunk 1 above)
Comment (sframba) [reply src/coreComponents/mesh/utilities/ComputationalGeometry.hpp:451]: Yes, for exemple here:
https://github.com/GEOS-DEV/GEOS/blob/9cf5ab5540ea296194363713f1bb3972dc618d53/src/coreComponents/physicsSolvers/wavePropagation/shared/PrecomputeSourcesAndReceiversKernel.hpp#L60
mesh coordinates (`nodeCoords`) are of type `WaveSolverBase::wsCoordType` (whihc is `real32`), while source and receiver coordinates are plain old `real64`. This is done for memory reasons on GPU, where having all mesh coordinates in `real64` can be too expensive

File: src/coreComponents/mesh/utilities/ComputationalGeometry.hpp (hunk 3 above)
Comment (sframba) [reply src/coreComponents/mesh/utilities/ComputationalGeometry.hpp]: Done

**Review by untereiner — APPROVED** [review]

**Comment by sframba** [comment]
@rrsettgast Is it good to go? If so, could you please add your review?

**Review by rrsettgast — APPROVED** [review]

**Pull request merged**